cookbooks/custom_formatter.rst

```diff
@@ -0,0 +1,464 @@
+           $formatterDefinition->addArgument('%paths.base%');
+           $formatterDefinition->addArgument($outputPrinterDefinition);
+
+           // tag the formatter as an "output.formatter", this way behat will add it to it's formatter list.
+           $formatterDefinition->addTag(OutputExtension::FORMATTER_TAG, ['priority' => 100]);
+       }
+
+       public function configure(ArrayNodeDefinition $builder): void { }
+
+       public function initialize(ExtensionManager $extensionManager): void { }
+
+       public function process(ContainerBuilder $container): void { }
+   }
+
+Create the formatter
+~~~~~~~~~~~~~~~~~~~~
+
+The formatter will listen to behat's events, and create output data depending on the type of event, the current state, etc.
+
+.. code:: php
+
+   <?php
+
+   namespace JDeniau\BehatReviewdogFormatter;
+
+   use Behat\Behat\EventDispatcher\Event\AfterStepTested;
+   use Behat\Behat\EventDispatcher\Event\StepTested;
+   use Behat\Behat\Tester\Result\ExecutedStepResult;
+   use Behat\Testwork\EventDispatcher\Event\BeforeExerciseCompleted;
+   use Behat\Testwork\Output\Formatter;
+   use Behat\Testwork\Output\Printer\OutputPrinter;
+
+   class ReviewdogFormatter implements Formatter
+   {
+       public function __construct(
+           private readonly string $pathsBase,
+           private readonly ReviewdogOutputPrinter $outputPrinter
+       ) {
+       }
+
+       /**
+        * setParameter will be called for each key given to the formatter in your behat.yml file. 
+        * We will see that later in the "integration".
+        * In our case, the only allowed parameter is a "file_name" that must be a string : the JSON file that we will write.
+        */
+       public function setParameter($name, $value): void
+       {
+           switch ($name) {
+               case 'file_name':
+                   if (!is_string($value)) {
+                       throw new \InvalidArgumentException('file_name must be a string');
+                   }
+
+                   $this->outputPrinter->setFileName($value);
+                   break;
+               default:
+                   throw new \Exception('Unknown parameter ' . $name);
+           }
+       }
+
+       /**
+        * I do not call this, so no need to define an implementation
+        */
+       public function getParameter($name) { }
+
+       /**
+        * Our formatter is a Symfony EventSubscriber. 
+        * This method tells behat where we want to "hook" in the process.
+        * Here we want to be called:
+        *   - at start, when the test is launched with the `BeforeExerciseCompleted::BEFORE` event,
+        *   - when a step has ended with the `StepTested::AFTER` event.
+        * 
+        * There is a lot of other that can be found here: https://github.com/Behat/Behat/tree/2a3832d9cb853a794af3a576f9e524ae460f3340/src/Behat/Testwork/EventDispatcher/Event
```

Comment:
I suggest not linking here - just to make sure people go and look for the most recent version in case the events change in future.

```suggestion
        * There are a lot of other events that can be found here in the Behat\Testwork\EventDispatcher\Event class
```